source/model/breadcrumbs.ts

```diff
@@ -0,0 +1,65 @@
+const typeManual = 'manual';
+const defaultLevel = 'info';
+
+export interface IBreadcrumb {
+        timestamp: number,
+        level: string,
+        type: string,
+        message: string,
+        attributes: object,
+}
+
+export class Breadcrumbs {
+        public static readonly attachmentName = 'backtrace-breadcrumbs';
+        private breadcrumbLimit: number = -1;
+        private _breadcrumbs: IBreadcrumb[] = [];
+        
+        constructor(breadcrumbLimit?: number) {
+                if (!breadcrumbLimit || breadcrumbLimit <= 0) {
+                        this.breadcrumbLimit = -1;
+                        return;
+                }
+                this.breadcrumbLimit = breadcrumbLimit;
+        }
+
+        public add(
+                message: string,
+                attributes={},
+                timestamp=this.getNowUnixTimestamp(),
+                level: string=defaultLevel, 
+                type: string=typeManual,
+                ) {
+                // breadcrumbs are disabled
+                if (this.breadcrumbLimit < 0) {
+                        return;
+                }
+                // if breadcrumbs array is full, purge oldest entry
+                if (this._breadcrumbs.length === this.breadcrumbLimit) {
```

Comment:
I think we need to run this condition in a while loop - just in case when multiple promises will generate breadcrumbs together.

if we will use the while loop to remove the oldest breadcrumbs, we will be always sure that we will have `this.breadcrumbLimit` limit - and we will handle correctly a case when two promises will add breadcrumbs at the same time. 

What do you think about this case?

Reply:
I don't actually think this is a necessary change due to single-threadedness of JavaScript, but the suggestion is a harmless change, so sure.